library/java/org/chromium/net/impl/CancelProofEnvoyStream.java

```diff
@@ -0,0 +1,242 @@
+package org.chromium.net.impl;
+
+import androidx.annotation.IntDef;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.IntUnaryOperator;
+import io.envoyproxy.envoymobile.engine.EnvoyHTTPStream;
+
+/**
+ * Consistency layer above the {@link EnvoyHTTPStream} preventing unwarranted Stream operations
+ * after a "cancel" operation. There are no "synchronized" - this is CAS based logic.
+ *
+ * <p>This contraption ensures that once a "cancel" operation is invoked, there will be no further
+ * operations allowed with the EnvoyHTTPStream - subsequent operations will be ignored silently.
+ * However, in the event that that one or more EnvoyHTTPStream operations are currently being
+ * executed, the "cancel" operation gets postponed: the last concurrent operation will invoke
+ * "cancel" at the end.
+ *
+ * <p>Instance of this class start with a state of "BUSY_STARTING". This ensure that if a cancel
+ * is invoked while the stream is being created, that cancel will be executed only once the stream
+ * is completely initialized. Doing otherwise leads to unpredictable outcomes.
+ */
+final class CancelProofEnvoyStream {
```

Comment:
Will this also be used for plain url requests?

Reply:
Yes. The same logic is already is already present in CronetUrlRequest, but spread around.  I need to clean that out. In fact, CronetUrlRequest needs a big cleanup.